### PR TITLE
Moodle course copy fix for Pages and Files

### PR DESCRIPTION
**Pull request by marcospri**
The fix for pages & files is very similar, tackling it in one PR.

## Testing

The course copy fixed rely on having information stored from the original course, to simulate that we are going to select Pages and files from the file picker in the original course and stop there, ie. just listing them for the side effect **but not saving the changes**.

- Log in in moodle as moodleteacher.

- "Select content" button in  [HTML - Group Assignment](https://hypothesisuniversity.moodlecloud.com/course/modedit.php?update=861&return=1)

- List Moodle Files
- List Moodle Pages

## Files

- [Launch the original assignment](https://hypothesisuniversity.moodlecloud.com/mod/lti/view.php?id=863&forceview=1)

Result in the following logging:

```
DEBUG [lms.views.api.moodle.files:52][MainThread] Via URL for file in the same course. moodle://file/course/11/url/https://hypothesisuniversity.moodlecloud.com/webservice/pluginfile.php/1046/mod_resource/content/1/dummy.pdf?forcedownload=1
```

- [Launch the equivalent assignment in the new course](https://hypothesisuniversity.moodlecloud.com/mod/lti/view.php?id=889&forceview=1)

which logs:

```
Via URL for page, found page in the new course. Document: 
moodle://file/course/11/url/https://hypothesisuniversity.moodlecloud.com/webservice/pluginfile.php/1046/mod_resource/content/1/dummy.pdf?forcedownload=1, 
course: 31, 
new page id: https://hypothesisuniversity.moodlecloud.com/webservice/pluginfile.php/1152/mod_folder/content/4/dummy.pdf?forcedownload=1
```

- Will launch [again the new assignment](https://hypothesisuniversity.moodlecloud.com/mod/lti/view.php?id=889&forceview=1) to get the new ID from the mapping

```
 Via URL for file already mapped for course copy. Document: moodle://file/course/11/url/https://hypothesisuniversity.moodlecloud.com/webservice/pluginfile.php/1046/mod_resource/content/1/dummy.pdf?forcedownload=1, course: 31, mapped file: https://hypothesisuniversity.moodlecloud.com/webservice/pluginfile.php/1152/mod_folder/content/4/dummy.pdf?forcedownload=1
```

## Pages

- [Launch the original assignment](https://hypothesisuniversity.moodlecloud.com/mod/lti/view.php?id=865)

Result in the following logging:

```
Via URL for page in the same course. moodle://page/course/11/page_id/860
```

- [Launch the equivalent assignment in the new course](https://hypothesisuniversity.moodlecloud.com/mod/lti/view.php?id=890&forceview=1)

```
Via URL for page, found page in the new course. Document: moodle://page/course/11/page_id/860, course: 31, new page id: 869
```

- Will launch [again the new assignment](https://hypothesisuniversity.moodlecloud.com/mod/lti/view.php?id=890&forceview=11) to get the new ID from the mapping

```
Via URL for page already mapped for course copy. Document: 
moodle://page/course/11/page_id/860, course: 31, mapped page_id: 869
```

### CODE REVIEW

File: lms/product/moodle/_plugin/misc.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from lms.product.plugin.misc import MiscPlugin
+
+
+class MoodleMiscPlugin(MiscPlugin):
+    deep_linking_prompt_for_title = False
```
Comment (marcospri) [inline lms/product/moodle/_plugin/misc.py:5]: I've been suffering this while testing, fixed it on this PR.

The main issue is while editing an existing assignment using Moodle we prompt for a title again, use the "Hypothesis Assignment" default and we overwrite the title of the assignment.

File: lms/services/moodle.py [hunk 2]
```diff
@@ -83,7 +86,18 @@ def list_files(self, course_id: int):
                         )
                     )
 
-        return self._construct_file_tree(course_id, files)
+        file_tree = self._construct_file_tree(course_id, files)
+        self._file_service.upsert(
```
Comment (marcospri) [inline lms/services/moodle.py:90]: We need to store files when we list them so we can use that info to fix course copy.

File: lms/services/moodle.py [hunk 3]
```diff
@@ -130,11 +144,21 @@ def list_pages(self, course_id: int):
                     file_node = {
                         "type": "Page",
                         "display_name": module["name"],
-                        "lms_id": f"moodle://page/course/{course_id}/page_id/{module['id']}",
+                        "lms_id": module["id"],
```
Comment (marcospri) [inline lms/services/moodle.py:147]: Pages do have IDs, use that instead of the full URL.

File: lms/services/moodle.py [hunk 4]
```diff
@@ -130,11 +144,21 @@ def list_pages(self, course_id: int):
...
                         "id": f"moodle://page/course/{course_id}/page_id/{module['id']}",
                     }
                     current_node["children"].append(file_node)
 
+        self._file_service.upsert(
```
Comment (marcospri) [inline lms/services/moodle.py:152]: Also store pages, same reasoning as for files.

File: lms/services/moodle.py [hunk 5]
```diff
@@ -179,8 +203,8 @@ def _construct_file_tree(course_id, files):
             file_node = {
                 "type": "File",
                 "display_name": path_components[-1],
-                "id": f"moodle://file/url/{file_data['url']}",
-                "lms_id": f"moodle://file/url/{file_data['url']}",
+                "id": f"moodle://file/course/{course_id}/url/{file_data['url']}",
```
Comment (marcospri) [inline lms/services/moodle.py:206]: And oversights for me in the first version of this. We need to store the course alongside the file info to make our lives easier for course copy.

File: lms/views/api/moodle/files.py [hunk 6]
```diff
@@ -1,12 +1,18 @@
...
 
 from lms.security import Permissions
+from lms.services.exceptions import FileNotFoundInCourse
 from lms.services.moodle import MoodleAPIClient
 from lms.views import helpers
 
-DOCUMENT_URL_REGEX = re.compile(r"moodle:\/\/file\/url\/(?P<url>.*)")
+LOG = getLogger(__name__)
+
+DOCUMENT_URL_REGEX = re.compile(
```
Comment (marcospri) [inline lms/views/api/moodle/files.py:13]: Including course/COURSE_ID in the regexp now.

File: lms/views/api/moodle/files.py [hunk 7]
```diff
@@ -29,12 +35,61 @@ def list_files(_context, request):
     permission=Permissions.API,
 )
 def via_url(_context, request):
-    token = request.find_service(MoodleAPIClient).token
+    course_copy_plugin = request.product.plugin.course_copy
```
Comment (marcospri) [inline lms/views/api/moodle/files.py:38]: The following code handles a few cases separately is very similar between pages and files. Also very similar to D2L.

File: lms/views/api/moodle/pages.py [hunk 8]
```diff
@@ -30,30 +35,73 @@ def list_pages(self):
...
+            current_course_id, raise_on_missing=True
         )
         assignment = self.request.find_service(name="assignment").get_assignment(
             self.request.lti_user.application_instance.tool_consumer_instance_guid,
             self.request.lti_user.lti.assignment_id,
         )
         document_url = assignment.document_url
-        _, document_page_id = self._parse_document_url(document_url)
+        document_course_id, document_page_id = self._parse_document_url(document_url)
+        effective_page_id = None
```
Comment (marcospri) [inline lms/views/api/moodle/pages.py:50]: Very similar to the files version.

Added a note about generalizing this to the board.

File: lms/views/api/moodle/files.py [hunk 9]
```diff
@@ -29,12 +35,61 @@ def list_files(_context, request):
...
+            )
+        # Store a mapping so we don't have to re-search next time.
+        LOG.debug(
+            "Via URL for page, found page in the new course. Document: %s, course: %s, new page id: %s",
+            document_url,
+            current_course_id,
+            found_file.lms_id,
+        )
+        course.set_mapped_file_id(document_file_id, found_file.lms_id)
+        effective_file_id = found_file.lms_id
```
Comment (acelaya) [inline lms/views/api/moodle/files.py:88]: This whole block feels a bit hard to follow. Perhaps extracting it into a private method with a couple early returns would help.

```python
def _resolve_file_id(document_course_id, document_file_id) -> str
    if current_course_id == document_course_id:
        # Not in a course copy scenario, use the IDs from the document_url
        LOG.debug("Via URL for file in the same course. %s", document_url)
        return document_file_id
    
    mapped_file_id = course.get_mapped_file_id(document_file_id)
    if mapped_file_id != document_file_id:
        LOG.debug(
            "Via URL for file already mapped for course copy. Document: %s, course: %s, mapped file: %s",
            document_url,
            current_course_id,
            mapped_file_id,
        )
        return mapped_file_id
    
    # In moodle course copy for files is easier to solve because we don't make
    # requests in the name of the user so we can fix it for all launches.
    # It won't only not succeed if the file doesn't have an equivalent file in the new course
    found_file = course_copy_plugin.find_matching_file_in_course(
        document_file_id, course.lms_id
    )
    if not found_file:
        LOG.debug(
            "Via URL for page, couldn't find page in the new course. Document: %s, course: %s.",
            document_url,
            current_course_id,
        )
        raise FileNotFoundInCourse(
            "moodle_file_not_found_in_course_instructor", document_url
        )
    
    # Store a mapping so we don't have to re-search next time.
    LOG.debug(
        "Via URL for page, found page in the new course. Document: %s, course: %s, new page id: %s",
        document_url,
        current_course_id,
        found_file.lms_id,
    )
    course.set_mapped_file_id(document_file_id, found_file.lms_id)
    return found_file.lms_id
```

And then in here you can do:

```python
document_course_id = DOCUMENT_URL_REGEX.search(document_url)["course_id"]
document_file_id = DOCUMENT_URL_REGEX.search(document_url)["url"]
effective_file_id = self._resolve_file_id(document_course_id, document_file_id)
```

File: lms/views/api/moodle/pages.py (hunk 8 above)
Comment (acelaya) [reply lms/views/api/moodle/pages.py:50]: I would also suggest extracting it into a private method as I did for files. I think it will help generalizing both of them later if we want.
